File: binary_search/binary_search.c
#include <stdio.h>
#include <stdlib.h>
#include "quicksort.h"

// arr is already sorted here
int binary_search_helper(int* arr, int low, int high, int val){

	if(high>=low){
		int mid = low + (high-low)/2;

		printf("%d  %d  %d  %d  %d\n",low, mid, high,arr[mid], val);
		if(arr[mid] == val){
			//int* i = &arr[mid];
			//i = &mid;
			return mid;
		}
		if(arr[mid]>val){
			return binary_search_helper(arr, low, mid-1, val);
		}else{
			return binary_search_helper(arr, mid+1, high, val); 
		}
	}else{
		return -1;
	}
}

int binary_search(int* arr, int low, int high, int val){
	quicksort(arr, low, high);
	return binary_search_helper(arr, low, high, val);;
}

File: algorithms/binary_search/binary_search.py
from random import randint

def partition(arr, low, high):
	'''Puts all elements of arr smaller than arr[high-1]
	to the left and all the larger to the right of the array
	'''
	pivot = arr[high]
	i = low -1 #Begins at 0

	for j in range(low, high):
		if arr[j] < pivot:
			i+=1
			arr[i], arr[j] = arr[j], arr[i]

	arr[i+1], arr[high] = arr[high], arr[i+1]

	return i+1

def quicksort(arr, low, high):
	if low < high:

		part = partition(arr, low, high)
		quicksort(arr, low, part-1)
		quicksort(arr, part+1, high)


def binary_search(arr, low, high, val):
	
	if high >= low:

		i = low + (high-low)//2

		if arr[i] == val:
			return i
		if arr[i] > val:
			return binary_search(arr, low, i-1, val)
		else:
			return binary_search(arr, i+1, high, val)
	else:
		return None


if __name__ == '__main__':

	# Tesitng random array
	to_sort = [randint(0,100) for i in range(50)]
	quicksort(to_sort, 0, len(to_sort)-1)
	print("Sorted array:\n{}".format(to_sort))

	val = 10
	s = binary_search(to_sort,0 ,len(to_sort)-1, val)

	if s is not None:
		print("Found element {} on index {}".format(val,s))



